src/libmeasurement_kit/engine.cpp

```diff
@@ -0,0 +1,480 @@
+// Part of Measurement Kit <https://measurement-kit.github.io/>.
+// Measurement Kit is free software under the BSD license. See AUTHORS
+// and LICENSE for more information on the copying conditions.
+
+#define MK_ENGINE_INTERNALS
+#include <measurement_kit/engine.h>
+
+#include <assert.h>
+#include <stdint.h>
+
+#include <algorithm>
+#include <atomic>
+#include <condition_variable>
+#include <deque>
+#include <exception>
+#include <future>
+#include <memory>
+#include <mutex>
+#include <set>
+#include <string>
+#include <sstream>
+#include <thread>
+#include <tuple>
+#include <type_traits>
+
+#include <measurement_kit/common/error.hpp>
+#include <measurement_kit/common/logger.hpp>
+#include <measurement_kit/common/nlohmann/json.hpp>
+#include <measurement_kit/common/reactor.hpp>
+#include <measurement_kit/common/shared_ptr.hpp>
+
+#include "src/libmeasurement_kit/nettests/runnable.hpp"
+
+namespace mk {
+namespace engine {
+
+// # Multi-thread stuff
+//
+// Comes first because it needs more careful handling.
+
+class Semaphore {
+  public:
+    Semaphore() = default;
+
+    void wait() {
+        std::unique_lock<std::mutex> lock{mutex_};
+        cond_.wait(lock, [this]() { return !active_; });
+        active_ = true;
+    }
+
+    void signal() {
```

Comment:
<Bikeshed>
I generally prefer the naming of `acquire` and `release` for semaphore (and find it more commonly used, for ex `golang`, `twisted`, etc.), but I don't have a strong opinion on this.
</Bikeshed>